### PR TITLE
Add conditions to FybrikModule

### PR DESCRIPTION
**Pull request by shlomitk1**
Signed-off-by: Shlomit Koyfman <shlomitk@il.ibm.com>

Fixes https://github.com/fybrik/fybrik/issues/966

This PR defines a Ready condition with information such as status, message and observed generation.

### CODE REVIEW

File: charts/fybrik-crd/templates/app.fybrik.io_fybrikmodules.yaml [hunk 1]
```diff
@@ -273,19 +273,32 @@ spec:
...
+                      description: Message contains the details of the current condition
+                      type: string
+                    observedGeneration:
+                      description: ObservedGeneration is the version of the resource
+                        for which the condition has been evaluated
+                      format: int64
+                      type: integer
+                    status:
+                      description: 'Status of the condition: true or false'
+                      type: string
```
Comment (roytman) [inline charts/fybrik-crd/templates/app.fybrik.io_fybrikmodules.yaml:293]: If Condition status can have only 3 values: "True", "False" and "Unknown", should we add a string validation, something like: "pattern: "False|True|Unknown"

Comment (shlomitk1) [reply charts/fybrik-crd/templates/app.fybrik.io_fybrikmodules.yaml:293]: For the best of my knowledge, we do not validate fields inside the status, only in spec.
I've used condition status similar to described in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md
I'll change the description to include Unknown.

Comment (roytman) [reply charts/fybrik-crd/templates/app.fybrik.io_fybrikmodules.yaml:293]: Actually, there is a better option, here is an example from `issuers.cert-manager.io`
```yaml
status:
  description: Status of the condition, one of (`True`, `False`, `Unknown`).
  enum:
  - "True"
  - "False"
  - Unknown
  type: string
```

File: manager/controllers/app/fybrikmodule_controller.go [hunk 2]
```diff
@@ -54,22 +55,27 @@ func (r *FybrikModuleReconciler) Reconcile(ctx context.Context, req ctrl.Request
 
 	observedStatus := moduleContext.Status.DeepCopy()
 	moduleVersion := moduleContext.GetGeneration()
+	if len(moduleContext.Status.Conditions) == 0 {
+		moduleContext.Status.Conditions = []api.Condition{{Type: api.ReadyCondition, Status: corev1.ConditionUnknown, ObservedGeneration: 0}}
+	}
 
 	// check if module has been validated before or if validated module is outdated
-	if string(moduleContext.Status.ValidModule) == "" || observedStatus.ValidatedGeneration != moduleVersion {
+	condition := moduleContext.Status.Conditions[0]
```
Comment (revit13) [inline manager/controllers/app/fybrikmodule_controller.go:63]: can `ReadyConditionIndex`  be used instead of 0?

Comment (revit13) [inline manager/controllers/app/fybrikmodule_controller.go:63]: where is the condition type is checked? it seems to be updated later without checking the type.

File: manager/apis/app/v1alpha1/condition_types.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	ErrorCondition ConditionType = "Error"
+	DenyCondition  ConditionType = "Deny"
+	ReadyCondition ConditionType = "Ready"
+)
+
+// Condition describes the state of a FybrikApplication at a certain point.
+type Condition struct {
+	// Type of the condition
+	Type ConditionType `json:"type"`
+	// Status of the condition: true or false
```
Comment (roytman) [inline manager/apis/app/v1alpha1/condition_types.go:24]: Should we add something like:
// +kubebuilder:validation:Optional
// +kubebuilder:validation:Enum=True,False,Unknown
 // +kubebuilder:default:=Unknown

File: manager/apis/app/v1alpha1/fybrikapplication_types.go [hunk 4]
```diff
@@ -102,38 +102,6 @@ const (
...
-	// ReadyCondition means that access to a dataset is granted
-	ReadyCondition ConditionType = "Ready"
-)
-
-// Condition describes the state of a FybrikApplication at a certain point.
-type Condition struct {
-	// Type of the condition
-	Type ConditionType `json:"type"`
-	// Status of the condition: true or false
-	Status corev1.ConditionStatus `json:"status"`
```
Comment (roytman) [inline manager/apis/app/v1alpha1/fybrikapplication_types.go:131]: as above

File: charts/fybrik-crd/templates/app.fybrik.io_fybrikmodules.yaml (hunk 1 above)
Comment (shlomitk1) [reply charts/fybrik-crd/templates/app.fybrik.io_fybrikmodules.yaml:293]: OK

File: manager/controllers/app/fybrikmodule_controller.go (hunk 2 above)
Comment (shlomitk1) [reply manager/controllers/app/fybrikmodule_controller.go]: Condition type is not checked. Conditions are assumed to be static, i.e. having the same indices in the array over the run.

File: manager/apis/app/v1alpha1/condition_types.go (hunk 3 above)
Comment (shlomitk1) [reply manager/apis/app/v1alpha1/condition_types.go]: Added

File: manager/controllers/app/fybrikmodule_controller.go (hunk 2 above)
Comment (revit13) [reply manager/controllers/app/fybrikmodule_controller.go]: OK. does it make sense to use `ReadyConditionIndex`  instead of 0?

**Review by revit13 — APPROVED** [review]

**Pull request merged**